### PR TITLE
hw1 for cse7345

### PR DESCRIPTION
**Pull request by ganktoby**
cse7345
lijie zhang
smu id: 47495426

### CODE REVIEW

File: ex01.ex [hunk 1]
```diff
@@ -60,7 +62,11 @@ defmodule Ex01 do
   # no explicit + operators in your function                          #
   #####################################################################
 
-  sum3a = your_anonymous_function(1, 2, 3)
+  sum3a = fn (a, b, c) ->
+    tempSum = sum2a.(a, b)
```
Comment (pragdave) [inline ex01.ex:66]: ## -1

Why did you use a temporary variable here and not in the next one?

In fact, this problem has an elegant solution using a pipeline:

~~~ elixir
  sum3a = fn a, b, c -> a |> sum2a.(b) |> sum2a.(c) end
~~~

File: ex02.ex [hunk 2]
```diff
@@ -41,7 +43,7 @@ defmodule Ex02 do
   # Do the same using the & syntax #
   ##################################
 
-  list2b = your_anonymous_function(1, 2)
+  list2b = &([(&1 + &2), (&1 - &2)])
```
Comment (pragdave) [inline ex02.ex:46]: No deductions, but I think the form without the unneeded parens is easier to read:

~~~ elixir
list2b = &[ &1 + &2, &1 - &2 ]
~~~

File: ex02.ex [hunk 3]
```diff
@@ -53,8 +55,10 @@ defmodule Ex02 do
   # if the first two elements of a list are equal                #
   ################################################################
 
-  first2equal = your_anonymous_function([])
-
+  first2equal = fn ([h|t]) ->
```
Comment (pragdave) [inline ex02.ex:58]: ## -2

You could use pattern matching in the original function to simplify this:

~~~ elixir
  first2equal = fn [ a, b | _ ]  -> a == b end
~~~

File: ex03.ex [hunk 4]
```diff
@@ -55,8 +55,22 @@ defmodule Ex03 do
 
   """
```
Comment (pragdave) [inline ex03.ex:57]: ## -2 

Wow. This is an interesting way of doing this. But it is unnecessarily complex. You could just do

~~~ elixir
  def odd_even([]), do: []
  def odd_even([ h | t ]), do: [ atom_for(h) | odd_even(t) ]

  defp atom_for(n) when Integer.is_odd(n),  do:  :odd
  defp atom_for(n) when Integer.is_even(n), do:  :even
~~~

File: ex04.ex [hunk 5]
```diff
@@ -75,7 +83,28 @@ defmodule Ex04 do
   return value will be the thing you have to manipulate.
   """
 
-  def even_odd . . . "your code"
+  def even_odd([h|t]) do
```
Comment (pragdave) [inline ex04.ex:86]: ## -4

There's a lot of duplication here, both in the code and in the fact that you process the list twice.

The comments made a strong suggestion that `reduce` would help:

~~~ elixir
  def even_odd(list) do
    list
    |> reverse()
    |> reduce({ [], [] }, &partition_even_odd/2)
  end

  defp partition_even_odd(x, { es, os }) when Integer.is_even(x), do: {[x | es], os}
  defp partition_even_odd(x, { es, os }) when Integer.is_odd(x),  do: {es, [x | os]}
~~~

**Comment by pragdave** [comment]
I've deducted 2 marks because you did your work in your master branch.

**Comment by ganktoby** [comment]
why I get point off for this problem?

Since mine is like sum3a = fn (a, b, c) -> sum2a.(sum2a.(a, b), c).
I am just thinking the nested function call would look bad, so I split into
two lines.
there is no logic condition stuff, why I get point off.

I understand your solution looks better, but mine is not that bad to get
point off, isn't it?

2017-09-19 11:51 GMT-05:00 Dave Thoms <notifications@github.com>:

**Comment by ganktoby** [comment]
The solution you provide is definitely correct and uncomplex. And it
actually is more common and forward sense. That was the first solution I
got for this question.
The reason why I decided to change to the current version,  is I trying to
coding without any logic condition, even the case, when(which is more
pattern match).
It might be a little bit complex, but it at least shows that I am thinking
more and trying to get used to the functional language.

I know it is a little bit complex, but do I have to get points off for this?

**Comment by ganktoby** [comment]
Sorry for the confusion.

My first reply was for ex01.ex, 1.3
My second reply was for ex03.ex, 3.1

I did not know all email gonna combined together.

Thanks, and sorry for wasting your time to read all my regrade request.

Best,
Lijie Zhang

2017-09-19 14:02 GMT-05:00 Lijie Zhang <zhanglijietoby@gmail.com>:

> The solution you provide is definitely correct and uncomplex. And it
> actually is more common and forward sense. That was the first solution I
> got for this question.
> The reason why I decided to change to the current version,  is I trying to
> coding without any logic condition, even the case, when(which is more
> pattern match).
> It might be a little bit complex, but it at least shows that I am thinking
> more and trying to get used to the functional language.
>
> I know it is a little bit complex, but do I have to get points off for
> this?
>

**Comment by pragdave** [comment]
I try to explain why I deduct marks in the comments in GitHub.
In your case, it was the temporary variable.
Regards

Dave Thomas
On Tue, Sep 19, 2017 at 1:51 PM, ganktoby <notifications@github.com> wrote:
why I get point off for this problem?

Since mine is like sum3a = fn (a, b, c) -> sum2a.(sum2a.(a, b), c).
I am just thinking the nested function call would look bad, so I split into
two lines.
there is no logic condition stuff, why I get point off.

I understand your solution looks better, but mine is not that bad to get
point off, isn't it?

2017-09-19 11:51 GMT-05:00 Dave Thoms <notifications@github.com>:

**Pull request closed without merging**